service/src/main/java/bio/terra/pipelines/service/PipelineRunsService.java

```diff
@@ -60,6 +64,84 @@ public PipelineRunsService(
     this.workspaceManagerService = workspaceManagerService;
   }
 
+  /**
+   * Prepare a new PipelineRun for a given pipeline and user-provided inputs. The caller provides a
+   * job uuid and any relevant pipeline inputs. Teaspoons calls WSM to generate a write-only SAS url
+   * for each file input, then persists the pipeline run to the database. Finally, Teaspoons returns
+   * the write SAS url to the user.
+   *
+   * <p>The user can subsequently use the SAS url(s) to upload their input file(s).
+   *
+   * @param pipeline the pipeline to run
+   * @param jobId the job uuid
+   * @param userId the user id
+   * @param userProvidedInputs the user-provided inputs
+   */
+  public Map<String, String> preparePipelineRun(
+      Pipeline pipeline, UUID jobId, String userId, Map<String, Object> userProvidedInputs) {
+
+    PipelinesEnum pipelineName = PipelinesEnum.valueOf(pipeline.getName().toUpperCase());
+
+    if (pipeline.getWorkspaceId() == null) {
+      throw new InternalServerErrorException("%s workspaceId not defined".formatted(pipelineName));
+    }
+
+    // check that we don't already have this jobId
+    if (Boolean.TRUE.equals(pipelineRunsRepository.existsByJobId(jobId))) {
+      throw new DuplicateObjectException(
+          "Duplicate jobId %s found. Please resubmit with a different jobId".formatted(jobId));
+    }
+
+    // define the path for each file input as the jobId + the default value
+    Map<String, String> userProvidedFileInputs =
+        pipeline.getPipelineInputDefinitions().stream()
+            .filter(PipelineInputDefinition::getUserProvided)
+            .filter(p -> p.getType().equals(PipelineInputTypesEnum.VCF))
+            .collect(
+                HashMap::new,
+                (m, p) ->
+                    m.put(
+                        p.getName(),
+                        "pipeline-run-user-input-files/%s/%s"
+                            .formatted(jobId, p.getDefaultValue())),
```

Comment:
this is definitely weird to have a user provided input where we just use the default value.  maybe we should make the user pass in a file name so that way we can treat it like ther user input that its meant to be?

Reply:
it's only the default value when it comes to naming. i wouldn't know how to enforce that name when they upload. e.g. what if they specify the input "my_first_file.vcf.gz" at prepareRun, but then upload "my_second_file.txt" to that path?